File: Modules/Network/iOS/OC/LSCHTTPTask.m
//
//  HTTPTask.m
//  Sample
//
//  Created by 冯鸿杰 on 2018/8/20.
//  Copyright © 2018年 vimfung. All rights reserved.
//

#import "LSCHTTPTask.h"
#import "LSCEncoding.h"
#import "LSCHTTPFile.h"

@interface LSCHTTPTask () <NSURLSessionDelegate,
                           NSURLSessionTaskDelegate,
                           NSURLSessionDataDelegate,
                           LSCExportTypeAnnotation>
{
@private
    
    //处理队列
    dispatch_queue_t _queue;
    //超时计时器
    NSTimer *_timer;
    //自身引用
    LSCHTTPTask *_selfReference;
    
    //请求对象
    NSMutableURLRequest *_request;
    //请求会话
    NSURLSession *_session;
    //请求会话任务
    NSURLSessionDataTask *_task;
    
    //回复数据
    NSMutableData *_responseData;
    //回复对象
    NSHTTPURLResponse *_response;
    //数据长度
    long long _contentLength;
    
    //回复回调
    LSCFunction *_resultHandler;
    //失败回调
    LSCFunction *_faultHandler;
    //上传回调
    LSCFunction *_uploadProgressHandler;
    //下载回调
    LSCFunction *_downloadProgressHandler;
}

@end

@implementation LSCHTTPTask

- (instancetype)init
{
    if (self = [super init])
    {
        self.timeout = 60;
    }
    return self;
}

- (void)setUrl:(NSString *)url
{
    if (!url)
    {
        _request = nil;
        return;
    }
    
    if (!_request)
    {
        _request = [[NSMutableURLRequest alloc] initWithURL:[NSURL URLWithString:url]];
    }
    else
    {
        _request.URL = [NSURL URLWithString:url];
    }
}

- (NSString *)url
{
    return _request.URL.absoluteString;
}


- (void)getWithResultHandler:(LSCFunction *)resultHandler
                faultHandler:(LSCFunction *)faultHandler
{
    _resultHandler = resultHandler;
    _faultHandler = faultHandler;
    
    _request.HTTPMethod = @"GET";
    [self _fillHttpHeaders];
    [self _sendRequest];
}

- (void)postWithParameters:(NSDictionary<NSString *, NSString *> *)parameters
             resultHandler:(LSCFunction *)resultHandler
              faultHandler:(LSCFunction *)faultHandler
{
    _resultHandler = resultHandler;
    _faultHandler = faultHandler;
    
    _request.HTTPMethod = @"POST";
    [self _fillHttpHeaders];
    
    if (parameters)
    {
        NSString *paramString = [self _parametersStringWithParameters:parameters];
        _request.HTTPBody = [paramString dataUsingEncoding:NSUTF8StringEncoding];
    }
    
    [self _sendRequest];
}

- (void)uploadWithFileParams:(NSDictionary<NSString *, LSCHTTPFile *> *)fileParams
                  parameters:(NSDictionary<NSString *, NSString *> *)parameters
               resultHandler:(LSCFunction *)resultHandler
                faultHandler:(LSCFunction *)faultHandler
             progressHandler:(LSCFunction *)progressHandler
{
    _resultHandler = resultHandler;
    _faultHandler = faultHandler;
    _uploadProgressHandler = progressHandler;
    
    _request.HTTPMethod = @"POST";
    [self _fillHttpHeaders];
    
    NSString *tmpBoundary = [NSString stringWithFormat:@"%u", arc4random() % (9999999 - 123400) + 123400];
    NSString *boundaryString = [NSString stringWithFormat:@"Boundary-%@", tmpBoundary];
    [_request addValue:[NSString stringWithFormat:@"multipart/form-data; boundary=%@", boundaryString]
    forHTTPHeaderField:@"Content-Type"];
    
    _request.HTTPBody = [self _multipartFormDataWithFileParams:fileParams
                                                    parameters:parameters
                                                boundaryString:boundaryString];
    
    [self _sendRequest];
}

- (void)downloadWithResultHandler:(LSCFunction *)resultHandler
                     faultHandler:(LSCFunction *)faultHandler
                  progressHandler:(LSCFunction *)progressHandler
{
    _resultHandler = resultHandler;
    _faultHandler = faultHandler;
    _downloadProgressHandler = progressHandler;
    
    _request.HTTPMethod = @"GET";
    [self _fillHttpHeaders];
    [self _sendRequest];
}

- (void)cancel
{
    __block typeof(self) theTask = self;
    [self _addQueueOperation:^{
        
        [theTask -> _task cancel];
        theTask -> _task = nil;
        
        [theTask -> _session invalidateAndCancel];
        theTask -> _session = nil;
        
        theTask -> _selfReference = nil;
        
    }];
}

#pragma mark - Private

/**
 填充HTTP请求头
 */
- (void)_fillHttpHeaders
{
    if (self.headers)
    {
        //设置HTTP头
        __block typeof(self) theTask = self;
        [self.headers enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
            
            [theTask -> _request setValue:obj forHTTPHeaderField:key];
            
        }];
    }
}

/**
 添加队列操作

 @param handler 操作处理
 */
- (void)_addQueueOperation:(void (^)(void))handler
{
    if (!_queue)
    {
        _queue = dispatch_queue_create("HTTPOptQueue", DISPATCH_QUEUE_SERIAL);
    }
    
    dispatch_async(_queue, handler);
}

/**
 发起请求
 */
- (void)_sendRequest
{
    [self cancel];
    
    _selfReference = self;
    
    __block typeof(self) theTask = self;
    [self _addQueueOperation:^{
        
        if (theTask -> _responseData == nil)
        {
            theTask -> _responseData = [NSMutableData data];
        }
        [theTask -> _responseData setData:[NSData data]];
        
        
        dispatch_sync(dispatch_get_main_queue(), ^{
            
            [theTask _addTimeoutListener];
            
            NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
            config.timeoutIntervalForRequest = theTask.timeout;
            
            theTask -> _session = [NSURLSession sessionWithConfiguration:config delegate:theTask delegateQueue:[NSOperationQueue mainQueue]];
            
            theTask -> _task = [theTask -> _session dataTaskWithRequest:theTask -> _request];
            [theTask -> _task resume];
            
        });
        
    }];
}

/**
 添加超时监听
 */
- (void)_addTimeoutListener
{
    [self _removeTimeoutListener];
    
    //加入超时检测
    _timer = [NSTimer scheduledTimerWithTimeInterval:self.timeout
                                              target:self
                                            selector:@selector(_timeoutHandler)
                                            userInfo:nil
                                             repeats:NO];
}

- (void)_timeoutHandler
{
    //取消请求
    [self cancel];
    [self _removeTimeoutListener];
}

/**
 移除超时检测
 */
- (void)_removeTimeoutListener
{
    [_timer invalidate];
    _timer = nil;
}

/**
 生成参数字符串

 @param parameters 参数集合
 @param encoding 字符串编码
 @return 参数字符串
 */
- (NSString *)_parametersStringWithParameters:(NSDictionary<NSString *, NSString *> *)parameters
{
    NSMutableString *requestString = [NSMutableString string];

    [parameters enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, NSString * _Nonnull obj, BOOL * _Nonnull stop) {
       
        [requestString appendFormat:@"%@=%@&",
         [LSCEncoding urlEncode:key],
         [LSCEncoding urlEncode:obj]];
        
    }];
    
    if (requestString.length > 0)
    {
        [requestString deleteCharactersInRange:NSMakeRange(requestString.length - 1, 1)];
    }
    
    return requestString;
}


/**
 将提交参数转换为MultipartForm组织形式数据

 @param fileParams 文件参数
 @param parameters 参数
 @param boundaryString 分隔字符串
 @param encoding 编码
 @return 转换后数据
 */
- (NSData *)_multipartFormDataWithFileParams:(NSDictionary<NSString *, LSCHTTPFile *> *)fileParams
                                  parameters:(NSDictionary<NSString *, NSString *> *)parameters
                                boundaryString:(NSString *)boundaryString
{
    NSMutableData *postData = [NSMutableData data];
    [postData appendData:[[NSString stringWithFormat:@"--%@\r\n", boundaryString] dataUsingEncoding:NSUTF8StringEncoding]];
    
    NSData *endItemBoundaryData = [[NSString stringWithFormat:@"\r\n--%@\r\n", boundaryString] dataUsingEncoding:NSUTF8StringEncoding];
    
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    NSArray<NSString *> *keyArray = [fileParams allKeys];
    NSInteger keyCount = [keyArray count] - 1;
    for (int i = 0; i < keyArray.count; i++)
    {
        NSString *fileKey = keyArray[i];
        LSCHTTPFile *file = fileParams[fileKey];
        
        if ([fileManager fileExistsAtPath:file.path])
        {
            [postData appendData:[[NSString stringWithFormat:
                                   @"Content-Disposition: form-data; name=\"%@\"; filename=\"%@\"\r\n",
                                   fileKey,
                                   [file.path lastPathComponent]]
                                  dataUsingEncoding:NSUTF8StringEncoding]];
            
            [postData appendData:[[NSString stringWithFormat:
                                   @"Content-Type: %@\r\n",
                                   file.mimeType]
                                  dataUsingEncoding:NSUTF8StringEncoding]];
            
            if (file.transferEncoding)
            {
                [postData appendData:[[NSString stringWithFormat:
                                       @"Content-Transfer-Encoding: %@\r\n",
                                       file.transferEncoding]
                                      dataUsingEncoding:NSUTF8StringEncoding]];
            }
            
            [postData appendData:[[NSString stringWithFormat:@"\r\n"] dataUsingEncoding:NSUTF8StringEncoding]];
            
            [postData appendData:[NSData dataWithContentsOfFile:file.path]];
            
            if (i < keyCount)
            {
                [postData appendData:endItemBoundaryData];
            }
        }
        
    }

    keyArray = [parameters allKeys];
    keyCount = [keyArray count] - 1;
    for (int i = 0; i < [keyArray count]; i++)
    {
        NSString *keyString = [keyArray objectAtIndex:i];
        NSString *value = parameters[keyString];
        
        [postData appendData:[[NSString stringWithFormat:
                               @"Content-Disposition: form-data; name=\"%@\"\r\n\r\n",
                               keyString]
                              dataUsingEncoding:NSUTF8StringEncoding]];
        [postData appendData:[value dataUsingEncoding:NSUTF8StringEncoding]];

        if (i < keyCount)
        {
            [postData appendData:endItemBoundaryData];
        }
    }

    [postData appendData:[[NSString stringWithFormat:
                           @"\r\n--%@--\r\n",
                           boundaryString]
                          dataUsingEncoding:NSUTF8StringEncoding]];

    return postData;
}

#pragma mark - LSCExportTypeAnnotation

+ (NSArray<NSString *> *)excludeExportInstanceMethods
{
    return @[NSStringFromSelector(@selector(URLSession:didReceiveChallenge:completionHandler:)),
             NSStringFromSelector(@selector(URLSession:task:didSendBodyData:totalBytesSent:totalBytesExpectedToSend:)),
             NSStringFromSelector(@selector(URLSession:dataTask:didReceiveData:)),
             NSStringFromSelector(@selector(URLSession:dataTask:didReceiveResponse:completionHandler:)),
             NSStringFromSelector(@selector(URLSession:task:didCompleteWithError:))];
}

#pragma mark - NSURLSessionTaskDelegate

- (void)URLSession:(NSURLSession *)session
              task:(NSURLSessionTask *)task
didReceiveChallenge:(NSURLAuthenticationChallenge *)challenge
 completionHandler:(void (^)(NSURLSessionAuthChallengeDisposition disposition, NSURLCredential * _Nullable credential))completionHandler
{
    NSURLCredential *credential = nil;
    NSURLSessionAuthChallengeDisposition disposition = NSURLSessionAuthChallengeRejectProtectionSpace;
    if ([challenge.protectionSpace.authenticationMethod isEqualToString:NSURLAuthenticationMethodServerTrust])
    {
        //如果是受信任证书则使用证书
        credential = [NSURLCredential credentialForTrust:challenge.protectionSpace.serverTrust];
        disposition = NSURLSessionAuthChallengeUseCredential;
    }
    
    if (completionHandler)
    {
        completionHandler(disposition, credential);
    }
}

- (void)URLSession:(NSURLSession *)session
              task:(NSURLSessionTask *)task
   didSendBodyData:(int64_t)bytesSent
    totalBytesSent:(int64_t)totalBytesSent
totalBytesExpectedToSend:(int64_t)totalBytesExpectedToSend
{
    if (_uploadProgressHandler)
    {
        NSArray<LSCValue *> *args = @[[LSCValue numberValue:@(totalBytesExpectedToSend)],
                                      [LSCValue numberValue:@(totalBytesSent)]];
        [_uploadProgressHandler invokeWithArguments:args];
    }
}

- (void)URLSession:(NSURLSession *)session
          dataTask:(NSURLSessionDataTask *)dataTask
    didReceiveData:(NSData *)data
{
    [_responseData appendData:data];
    
    if (_downloadProgressHandler)
    {
        NSArray<LSCValue *> *args = @[[LSCValue numberValue:@(_contentLength)],
                                      [LSCValue numberValue:@(_responseData.length)]];
        [_downloadProgressHandler invokeWithArguments:args];
    }
}

- (void)URLSession:(NSURLSession *)session
          dataTask:(NSURLSessionDataTask *)dataTask
didReceiveResponse:(NSURLResponse *)response
 completionHandler:(void (^)(NSURLSessionResponseDisposition disposition))completionHandler
{
    _response = (NSHTTPURLResponse *)response;
    _contentLength = [[_response allHeaderFields][@"Content-Length"] longLongValue];
    
    if (completionHandler)
    {
        completionHandler(NSURLSessionResponseAllow);
    }
}

- (void)URLSession:(NSURLSession *)session
              task:(NSURLSessionTask *)task
didCompleteWithError:(nullable NSError *)error
{
    //取消超时计时器
    [self _removeTimeoutListener];
    
    if (error)
    {
        //请求错误
        if (_faultHandler)
        {
            [_faultHandler invokeWithArguments:@[[LSCValue stringValue:error.localizedFailureReason]]];
        }
        
        [_session finishTasksAndInvalidate];
        
        _task = nil;
        _session = nil;
        _selfReference = nil;
    }
    else
    {
        //请求成功
        
        if (_resultHandler)
        {
            NSArray<LSCValue *> *args = @[[LSCValue integerValue:_response.statusCode],
                                          [LSCValue dataValue:_responseData]];
            [_resultHandler invokeWithArguments:args];
        }
        
        [_session finishTasksAndInvalidate];
        _task = nil;
        _session = nil;
        _selfReference = nil;
    }
}

@end

File: Modules/Network/iOS/Swift/LSCHTTPTask.swift
//
//  LSCHTTPTask.swift
//  Sample
//
//  Created by 冯鸿杰 on 2018/8/21.
//  Copyright © 2018年 vimfung. All rights reserved.
//

#if os(iOS)

import LuaScriptCore_iOS_Swift

#elseif os(OSX)

import LuaScriptCore_OSX_Swift

#endif

/// HTTP任务
@objc(LSCHTTPTask)
class LSCHTTPTask: NSObject, LuaExportType
{
    var _request : URLRequest?;
    var _resultHandler : LuaFunction?;
    var _faultHandler : LuaFunction?;
    var _uploadProgressHandler : LuaFunction?;
    var _downloadProgressHandler : LuaFunction?;
    
    var _session : URLSession?;
    var _task : URLSessionDataTask?;
    
    var _responseData : Data?;
    var _response : HTTPURLResponse?;
    var _contentLength : UInt64 = 0;
    
    var _optQueue : DispatchQueue?;
    var _selfReference : LSCHTTPTask?;
    var _timer : Timer?;
    
    /// 请求地址
    @objc var url : String?
    {
        set
        {
            if newValue == nil
            {
                _request = nil;
                return;
            }
            
            let url : URL? = URL(string: newValue!);
            if url == nil
            {
                _request = nil;
                return;
            }
            
            if _request == nil
            {
                _request = URLRequest(url: url!);
            }
            else
            {
                _request?.url = url;
            }
        }
        get
        {
            return _request?.url?.absoluteString;
        }
    }
    
    
    /// 请求头集合
    @objc var headers : Dictionary<String, String>?;
    
    /// 超时时间
    @objc var timeout : TimeInterval = 60;
    
    
    /// 发起GET请求
    ///
    /// - Parameters:
    ///   - reusltHandler: 返回回调
    ///   - faultHandler: 失败回调
    @objc func get(resultHandler : LuaFunction?, faultHandler : LuaFunction?) -> Void
    {
        _resultHandler = resultHandler;
        _faultHandler = faultHandler;

        _request?.httpMethod = "GET";
        
        _sendRequest();
    }
    
    
    /// 发起POST请求
    ///
    /// - Parameters:
    ///   - parameters: 请求参数
    ///   - resultHandler: 返回回调
    ///   - faultHandler: 失败回调
    @objc func post(parameters : Dictionary<String, String>?,
                    resultHandler : LuaFunction?,
                    faultHandler : LuaFunction?) -> Void
    {
        _resultHandler = resultHandler;
        _faultHandler = faultHandler;
        
        _request?.httpMethod = "POST";
        
        if parameters != nil
        {
            let paramStr = _paramsString(params: parameters!);
            _request?.httpBody = paramStr.data(using: .utf8);
        }
        
        _sendRequest();
    }
    
    
    /// 上传文件
    ///
    /// - Parameters:
    ///   - fileParams: 文件参数集合
    ///   - parameters: 请求参数集合
    ///   - resultHandler: 返回回调
    ///   - faultHandler: 失败回调
    ///   - progressHandler: 上传进度回调
    @objc func upload(fileParams : Dictionary<String, LSCHTTPFile>,
                      parameters : Dictionary<String, String>?,
                      resultHandler : LuaFunction?,
                      faultHandler : LuaFunction?,
                      progressHandler : LuaFunction?) -> Void
    {
        _resultHandler = resultHandler;
        _faultHandler = faultHandler;
        _uploadProgressHandler = progressHandler;
        
        _request?.httpMethod = "POST";
        
        let tmpBoundary : String = "\(arc4random() % (9999999 - 123400) + 123400)";
        let boundaryString : String = "Boundary-\(tmpBoundary)";
        
        _request?.addValue("multipart/form-data; boundary=\(boundaryString)",
            forHTTPHeaderField: "Content-Type");
        
        _request?.httpBody = _multipartForm(fileParams: fileParams,
                                            parameters: parameters,
                                            boundaryString: boundaryString);
        
        _sendRequest();
    }
    
    
    /// 下载文件
    ///
    /// - Parameters:
    ///   - resultHandler: 返回回调
    ///   - faultHandler: 失败回调
    ///   - progressHandler: 下载进度回调
    @objc func download(resultHandler : LuaFunction?,
                        faultHandler : LuaFunction?,
                        progressHandler : LuaFunction?) -> Void
    {
        _resultHandler = resultHandler;
        _faultHandler = faultHandler;
        _downloadProgressHandler = progressHandler;
        
        _request?.httpMethod = "GET";
        
        _sendRequest();
    }
    
    
    /// 取消请求
    @objc func cancel() -> Void
    {
        _addQueueOperation {
            [weak self] in
            
            self?._task?.cancel();
            self?._task = nil;
            
            self?._session?.invalidateAndCancel();
            self?._session = nil;
            
            self?._selfReference = nil;
        };
    }
    
    
    /// 添加队列操作
    ///
    /// - Parameter handler: 处理器
    func _addQueueOperation(handler: @escaping () -> Void) -> Void
    {
        if _optQueue == nil
        {
            _optQueue = DispatchQueue(label:"HTTPOptQueue");
        }

        _optQueue?.async(execute: handler);
    }
    
    /// 发送请求
    func _sendRequest() -> Void
    {
        cancel();
        
        _selfReference = self;
        
        _addQueueOperation {
            [weak self] in
            
            self?._responseData = Data();
        
            DispatchQueue.main.async {
                
                self?._addTimeoutListener();
                
                let config : URLSessionConfiguration = URLSessionConfiguration.default;
                config.timeoutIntervalForRequest = (self?.timeout)!;
                
                self?._session = URLSession(configuration: config,
                                            delegate: self,
                                            delegateQueue: OperationQueue.main);
                self?._task = self?._session?.dataTask(with: (self?._request)!);
                self?._task?.resume();
            };
        }
    }
    
    
    /// 添加超时监听
    func _addTimeoutListener () -> Void
    {
        _removeTimeoutListener();
        _timer = Timer(timeInterval: timeout,
                       target: self,
                       selector: #selector(_timeoutHandler(timer:)),
                       userInfo: nil,
                       repeats: false);
    }
    
    
    /// 超时处理器
    ///
    /// - Parameter timer: 定时器
    @objc private func _timeoutHandler (timer : Timer) -> Void
    {
        cancel();
        _removeTimeoutListener();
    }
    
    
    /// 移除超时监听
    func _removeTimeoutListener () -> Void
    {
        _timer?.invalidate();
        _timer = nil;
    }
    
    
    /// 转换参数为参数字符串
    ///
    /// - Parameter params: 参数集合
    /// - Returns: 参数字符串
    func _paramsString(params : Dictionary<String, String>) -> String
    {
        var requestString : String = "";

        for (key, value) in params
        {
            let encodeKey : String = LSCEncoding.urlEncode(text: key)!;
            let encodeValue : String = LSCEncoding.urlEncode(text: value)!;
            requestString.append("\(encodeKey)=\(encodeValue)&");
        }
        
        if requestString.count > 0
        {
            //移除最后一个字符
            requestString.remove(at: requestString.index(before: requestString.endIndex));
        }
        
        return requestString;
    }
    
    /// 组织Multipart表单数据
    ///
    /// - Parameters:
    ///   - fileParams: 文件参数集合
    ///   - parameters: 参数集合
    ///   - boundaryString: 分隔符
    /// - Returns: 表单数据
    func _multipartForm(fileParams : Dictionary<String, LSCHTTPFile>,
                        parameters : Dictionary<String, String>?,
                        boundaryString : String) -> Data?
    {
        var postData : Data = Data();
        postData.append("--\(boundaryString)\r\n".data(using: .utf8)!);
        
        let endItemBoundaryData : Data = "\r\n--\(boundaryString)\r\n".data(using: .utf8)!;
        
        let fileManager : FileManager = FileManager.default;
        
        var index : Int = 0;
        for (key, value) in fileParams
        {
            if fileManager.fileExists(atPath: value.path!)
            {
                do
                {
                    let fileUrl : URL = URL(fileURLWithPath: value.path!);
                    let fileData = try Data(contentsOf: fileUrl);
                    
                    let itemDescData : Data? = "Content-Disposition: form-data; name=\"\(key)\"; filename=\"\(fileUrl.lastPathComponent)\"\r\n".data(using: .utf8);
                    postData.append(itemDescData!);
                    
                    let mimeType : String = value.mimeType;
                    let contentTypeData : Data? = "Content-Type: \(mimeType)\r\n".data(using: .utf8);
                    postData.append(contentTypeData!);
                    
                    if value.transferEncoding != nil
                    {
                        let transEncoding : String = value.transferEncoding!;
                        let tranfEncodingData : Data? = "Content-Transfer-Encoding: \(transEncoding)\r\n".data(using: .utf8);
                        postData.append(tranfEncodingData!);
                    }
                    
                    postData.append("\r\n".data(using: .utf8)!);
                    
                    postData.append(fileData);
                    
                    if index < fileParams.count - 1
                    {
                        postData.append(endItemBoundaryData);
                    }
                }
                catch
                {
                    
                }
            }
            
            index += 1;
        }
        
        if parameters != nil
        {
            index = 0;
            for (key, value) in parameters!
            {
                let itemDescData : Data? = "Content-Disposition: form-data; name=\"%\(key)\"\r\n\r\n".data(using: .utf8);
                postData.append(itemDescData!);
                
                let itemValueData : Data? = value.data(using: .utf8);
                postData.append(itemValueData!);
                
                if index < fileParams.count - 1
                {
                    postData.append(endItemBoundaryData);
                }
                
                index += 1;
            }
        }
        
        let endData : Data? = "\r\n--\(boundaryString)--\r\n".data(using: .utf8);
        postData.append(endData!);
        
        return postData;
    }
}

extension LSCHTTPTask : URLSessionDelegate
{
    
}

extension LSCHTTPTask : URLSessionTaskDelegate
{
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
    {
        var credential : URLCredential? = nil;
        var disposition : URLSession.AuthChallengeDisposition = .rejectProtectionSpace;
        
        if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust
        {
            credential = URLCredential(trust: challenge.protectionSpace.serverTrust!);
            disposition = .useCredential;
        }
        
        completionHandler(disposition, credential);
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64) {
        
        if _uploadProgressHandler != nil
        {
            let args : Array<LuaValue> = [LuaValue(intValue: totalBytesExpectedToSend), LuaValue(intValue: totalBytesSent)];
            _ = _uploadProgressHandler?.invoke(arguments: args);
        }

    }
    
}

extension LSCHTTPTask : URLSessionDataDelegate
{
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data)
    {
        _responseData?.append(data);
        
        if _downloadProgressHandler != nil
        {
            let args : Array<LuaValue> = [LuaValue(intValue: _contentLength), LuaValue(intValue: (_responseData?.count)!)];
            _ = _downloadProgressHandler?.invoke(arguments: args);
        }
    }
    
    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive response: URLResponse, completionHandler: @escaping (URLSession.ResponseDisposition) -> Void)
    {
        _response = response as? HTTPURLResponse;
        
        let contentLength : String? = _response?.allHeaderFields["Content-Length"] as? String;
        _contentLength = UInt64(contentLength!)!;

        completionHandler(.allow);
    }
    
    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?)
    {
        _removeTimeoutListener();
        
        if error != nil
        {
            if _faultHandler != nil
            {
                _ = _faultHandler?.invoke(arguments: [LuaValue(stringValue: (error?.localizedDescription)!)]);
            }
        }
        else
        {
            if _resultHandler != nil
            {
                let args : Array<LuaValue> = [LuaValue(intValue: (_response?.statusCode)!), LuaValue(dataValue: _responseData!)];
                _ = _resultHandler?.invoke(arguments: args);
            }
        }
        
        _session?.finishTasksAndInvalidate();
        
        _task = nil;
        _session = nil;
        _selfReference = nil;
    }
}
